src/main/java/com/jossemargt/cookietwist/signature/SignatureHasher.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * The MIT License
+ *
+ * Copyright (C) 2018 Jonnatan Jossemar Cordero
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.jossemargt.cookietwist.signature;
+
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidKeyException;
+import java.util.Formatter;
+import java.util.Objects;
+
+import javax.crypto.Mac;
+
+/**
+ * The abstract Class SignatureHasher handles the cryptographic signature
+ * calculation regardless the simmtric algorithm to be use, which has to be
+ * provided by any derivated class in the init method.
+ */
+public abstract class SignatureHasher {
+
+    /** The hmac secret key byte array. */
+    protected byte[] hmacSecretKey;
+
+    /** The hasher instance from {@link Mac}. */
+    protected Mac hasher;
+
+    /**
+     * Instantiates a new signature hasher with a secret key String and a
+     * un-initialized {@link Mac} instance.
+     *
+     * @param secret
+     *            the secret String UTF-8 encoded
+     */
+    public SignatureHasher(String secret) {
+        Objects.requireNonNull(secret, "Null secret key");
+        this.hmacSecretKey = secret.getBytes(StandardCharsets.UTF_8);
+    }
+
+    /**
+     * Initializes the {@link Mac} instance with an specific symmetric signature
+     * algorithm and the secrect key.
+     *
+     * @throws InvalidKeyException
+     *             the invalid key exception is thrown when a the {@link Mac} is
+     *             initialized with an invalid secret key.
+     */
+    public abstract void init() throws InvalidKeyException;
+
+    /**
+     * To hex string, transforms a byte array to its hexadecimal String
+     * representation byte per byte.
+     *
+     * @param bytes
+     *            the byte array to be formatted
+     * @return the formmated string
+     */
+    private String toHexString(byte[] bytes) {
```

Comment:
To make the `hasher` more cohesive, what about renaming this method to `convertHasherToHexString` and, instead of passing a `byte[]` just use the hasher that you have as a field of this class?

Reply:
Hmm... I going to do it, but to not continue in my blindless, do you have any article or book section that I can read more about it?

Reply:
After some google search, I'm still not sure if this change improves the overall behavior, because by referencing the ` hasher` in the `toHexString` method context could lead to an unwanted side effect which is resetting the HMAC engine without knowing if that should be done in the current state. So I'm going to leave it as is right know, but feel free to reopen this topic if you find something that I didn't see.

As always thanks for your reviews and point of view, I found them very helpful.